dialogue-core/src/main/java/com/palantir/dialogue/core/ConcurrencyLimitedChannel.java

```diff
@@ -0,0 +1,113 @@
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * A channel that monitors the successes and failures of requests in order to determine the number of concurrent
+ * requests allowed to a particular channel. If the channel's concurrency limit has been reached, the maybeCreateCall
+ * method returns empty
+ */
+final class ConcurrencyLimitedChannel implements LimitedChannel {
+    private static final Void NO_CONTEXT = null;
+
+    private final Channel delegate;
+    private final LoadingCache<Endpoint, Limiter<Void>> limiters;
+
+    @VisibleForTesting
+    ConcurrencyLimitedChannel(Channel delegate, Supplier<Limiter<Void>> limiterSupplier) {
+        this.delegate = delegate;
+        this.limiters = Caffeine.newBuilder()
+                .expireAfterAccess(Duration.ofMinutes(5))
+                .build(key -> limiterSupplier.get());
+    }
+
+    static ConcurrencyLimitedChannel create(Channel delegate) {
+        return new ConcurrencyLimitedChannel(delegate, ConcurrencyLimitedChannel::createLimiter);
+    }
+
+    private static Limiter<Void> createLimiter() {
+        AIMDLimit aimdLimit = AIMDLimit.newBuilder()
+                // Explicitly set values to prevent library changes from breaking us
+                .initialLimit(20)
+                .minLimit(1)
+                .maxLimit(200)
+                .backoffRatio(0.9)
+                // Don't count slow calls as a sign of the server being overloaded
+                .timeout(Long.MAX_VALUE, TimeUnit.DAYS)
+                .build();
+        return SimpleLimiter.newBuilder()
+                .limit(WindowedLimit.newBuilder().build(aimdLimit))
+                .build();
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeCreateCall(Endpoint endpoint, Request request) {
+        return limiters.get(endpoint).acquire(NO_CONTEXT).map(listener -> {
+            ListenableFuture<Response> call = delegate.createCall(endpoint, request);
+            Futures.addCallback(call, new LimiterCallback(listener), MoreExecutors.directExecutor());
+            return call;
+        });
+    }
+
+    /**
+     * Signals back to the {@link Limiter} whether or not the request was successfully handled.
+     */
+    private static final class LimiterCallback implements FutureCallback<Response> {
+        private static final ImmutableSet<Integer> DROP_CODES = ImmutableSet.of(429, 503);
+
+        private final Limiter.Listener listener;
+
+        private LimiterCallback(Limiter.Listener listener) {
+            this.listener = listener;
+        }
+
+        @Override
+        public void onSuccess(Response result) {
+            if (DROP_CODES.contains(result.code())) {
+                listener.onDropped();
+            } else {
+                listener.onSuccess();
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable throwable) {
+            listener.onIgnore();
```

Comment:
what would a channel timeout or a connection refused exception mean for the limiter?

Reply:
this matches today's behavior, which is that timeouts + connection refused do not actually adjust the limiter